050-Kafka-Node-Pools.md

```diff
@@ -0,0 +1,832 @@
+  annotations:
+    strimzi.io/node-pools: enabled
+spec:
+  kafka:
+    version: 3.4.0
+    listeners:
+      - name: plain
+        port: 9092
+        type: internal
+        tls: false
+      - name: tls
+        port: 9093
+        type: internal
+        tls: true
+    config:
+      offsets.topic.replication.factor: 3
+      transaction.state.log.replication.factor: 3
+      transaction.state.log.min.isr: 2
+      default.replication.factor: 3
+      min.insync.replicas: 2
+      inter.broker.protocol.version: "3.3"
+  # ...
+---
+
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: big-nodes
+  namespace: my-namespace
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  resources:
+    requests:
+      memory: 64Gi
+      cpu: 8
+    limits:
+      memory: 64Gi
+      cpu: 12
+  storage:
+    type: jbod
+    volumes:
+    - id: 0
+      type: persistent-claim
+      size: 1000Gi
+      class: fast-storage
+    - id: 1
+      type: persistent-claim
+      size: 1000Gi
+      class: fast-storage
+---
+
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: small-nodes
+  namespace: my-namespace
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  resources:
+    requests:
+      memory: 4Gi
+      cpu: 1
+    limits:
+      memory: 4Gi
+      cpu: 2
+  storage:
+    type: jbod
+    volumes:
+    - id: 0
+      type: persistent-claim
+      size: 100Gi
+      class: regular-storage
+```
+
+#### Protecting `KafkaNodePool` resources from being shared between multiple Kafka clusters
+
+The `strimzi.io/cluster` label should guarantee that the node pool will match only one Kafka cluster.
+Additional protection will be made against undesired moves between different Kafka clusters when the `strimzi.io/cluster` label is modified.
+The `.status` section of the `KafkaNodePool` resource will contain the Kafka cluster ID of the cluster it belongs to.
+For example:
+
+```yaml=
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: my-pool
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  # ...
+status:
+  # ...
+  clusterId: cABpUKeDTqi7q_XOtBMgUv
+```
+
+The cluster ID is already part of the `Kafka` CR `.status` section.
+The operator will check the cluster IDs in every reconciliation and if the cluster IDs don't match between the `Kafka` and `KafkaNodePool` CRs, it will expect that there is some kind of an misconfiguration.
+It will throw an `InvalidResourceException` and wait for the user to fix the issue.
+
+### Scaling the node pools
+
+The `KafkaNodePool` resource will also support the `scale` sub-resource.
+Support for the `scale` sub-resource should allow integration with Kubernetes Horizontal Pod Autoscaler.
+That on its own doesn't really provide fully functional Kafka auto-scaling mechanism.
+For example, if the load is coming from a single partition which is handling too many messages, adding more nodes will not help.
```

Comment:
Should we keep the scale sub-resource support out of this proposal? 

I think that use cases for the HPA integration are not clear, instead a vertical autoscaler may help with the single partition issue and rebalances. The auto scaling feature deserves its own proposal and it may include the NodePool scale sub-resource improvement. In this proposal we could simply mention that possibility.

Reply:
I think you are mixing different things. Scale subresource does not mean autoscaling on its own. It means the ability to scale the resource independently without understanding its structure which is the basic building block of autoscaling solution. We already support the scale subresource almost everywhere apart from the `Kafka` CR because that has 2 replicas fields and `KafkaTopic` as partition count cnnot be decreased. So I think having the scale subresource is in sync with the rest of Strimzi.

Reply:
OK, but if scale support is the default, why do we need to mention HPA and auto-scaling in this context? I think this is confusing.

Reply:
The scale subresource is a key building block for any auto-scaling solution and auto-scaling is something what was asked for by the users in the past. It also plays an important role for the motivation why to use separate custom resource for the node pools. That is why it is worth mentioning in the proposal. But at the same time, I feel like it is important to clarify that this does not mean that this proposal somehow solves auto-scaling. It just respects it as one of the possible future improvements which can build on top of this.

At the same time, the scale subresource is cheap implementation wise. So if nobody uses it because there is no feasible way to autoscale Kafka, then I don't think we care.